### PR TITLE
Add support for "execute immediate" for python files

### PR DESCRIPTION
**Pull request by sfc-gh-turbaszek**
### Pre-review checklist
   * [ ] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [ ] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually.
   * [ ] I've confirmed that my changes are up-to-date with the target branch.
   * [ ] I've described my changes in the release notes.
   * [ ] I've described my changes in the section below.

### Changes description
...

### CODE REVIEW

File: src/snowflake/cli/api/constants.py [hunk 1]
```diff
@@ -20,6 +20,7 @@
 
 TEMPLATES_PATH = Path(__file__).parent.parent / "templates"
 DEPLOYMENT_STAGE = "deployments"
+PYTHON_3_12 = (3, 12)
```
Comment (sfc-gh-turbaszek) [inline src/snowflake/cli/api/constants.py:23]: @sfc-gh-jsikorski

File: RELEASE-NOTES.md [hunk 2]
```diff
@@ -80,9 +80,10 @@
 * Added templating support for project definition file.
   * Template variables can now be used anywhere in the project definition file.
 * Added `--default` flag to `snow connection add` commands allowing users to mark the new connection as default.
+* Support for Python remote execution via `snow stage execute` and `snow git execute` similar to existing EXECUTE IMMEDIATE support.
```
Comment (sfc-gh-astus) [inline RELEASE-NOTES.md:83]: Move to 2.7.0

File: RELEASE-NOTES.md [hunk 3]
```diff
@@ -80,9 +80,10 @@
...
 
 ## Fixes and improvements
-* Fixed error handling for malformatted `config.toml`
+* Fixed error handling for malformed `config.toml`
```
Comment (sfc-gh-astus) [inline RELEASE-NOTES.md:86]: Move to 2.7.0

File: tests/git/__snapshots__/test_git_commands.ambr [hunk 4]
```diff
@@ -2,6 +2,8 @@
 # name: test_execute[@db.schema.repo/branches/main/-@db.schema.repo/branches/main/-expected_files2]
   '''
   SUCCESS - @db.schema.repo/branches/main/s1.sql
+  SUCCESS - @db.schema.repo/branches/main/s1.sql
+  SUCCESS - @db.schema.repo/branches/main/a/s3.sql
```
Comment (sfc-gh-astus) [inline tests/git/__snapshots__/test_git_commands.ambr:6]: Printing all files twice 😮

Comment (sfc-gh-turbaszek) [reply tests/git/__snapshots__/test_git_commands.ambr:6]: Good catch!

File: src/snowflake/cli/plugins/stage/manager.py [hunk 5]
```diff
@@ -335,38 +377,144 @@ def _filter_supported_files(files: List[str]) -> List[str]:
...
+        if sys.version_info >= PYTHON_3_12:
+            raise ClickException(
+                f"Executing python files is not supported in Python >= 3.12. Current version: {sys.version}"
+            )
+
+        from snowflake.snowpark.functions import sproc
+
+        self.snowpark_session.add_packages("snowflake-snowpark-python")
+        self.snowpark_session.add_packages("snowflake.core")
+        requirements = self._check_for_requirements_file(stage_path_parts)
```
Comment (sfc-gh-mraba) [inline src/snowflake/cli/plugins/stage/manager.py:479]: Should we skip empty lines or `snowpark.add_package()` can handle them ?

`requirements.txt`

```
a==1.0
b==2.0

# some comment in file
c==3.0
```

**Review by sfc-gh-astus — DISMISSED** [review]

File: src/snowflake/cli/plugins/stage/manager.py (hunk 5 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/plugins/stage/manager.py:490]: Fixed

File: src/snowflake/cli/api/sql_execution.py [hunk 6]
```diff
@@ -41,12 +41,22 @@
...
 
+    @property
+    def snowpark_session(self):
+        if not self._snowpark_session:
+            from snowflake.snowpark.session import Session
+
+            self._snowpark_session = Session.builder.configs(
+                {"connection": self._conn}
+            ).create()
+        return self._snowpark_session
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/api/sql_execution.py:58]: Why you open second connection?

File: src/snowflake/cli/plugins/stage/manager.py [hunk 7]
```diff
@@ -343,38 +386,145 @@ def _filter_supported_files(files: List[str]) -> List[str]:
...
+            if req_file in req_files_on_stage:
+                requirements_file = req_file
+                break
+
+        # If we haven't found any matching requirements
+        if requirements_file is None:
+            return []
+
+        # req_file at this moment is the first found requirements file
+        with TemporaryDirectory() as tmp_dir:
```
Comment (sfc-gh-astus) [inline src/snowflake/cli/plugins/stage/manager.py:467]: `with SecurirePath.temporary_directory() as tmp_dir:`

File: src/snowflake/cli/api/sql_execution.py (hunk 6 above)
Comment (sfc-gh-turbaszek) [reply src/snowflake/cli/api/sql_execution.py:58]: It's the same connection, but wrapped around Snowpark util

**Review by sfc-gh-mraba — APPROVED** [review]

**Pull request merged**